### DISCUSSION TITLE
Problem with messages received from whtasapp web

### DISCUSSION THREAD

**Discussion by ricardopereyra** [tulir/whatsmeow#568] [WhatsApp protocol Q&A]
Hello. I have a problem when a message is received from whatsapp web: 18:28:04.683 [Client ERROR] Event handler panicked while handling a *events.Message: message recipient must be a user JID with no device part.

Debugging I found that on a cellphone message v.Info.Sender is xxxxxxxxxx@s.whatsapp.net, but on a web whatsapp message v.Info.Sender is xxxxxxxxxxxxx:xx@s.whatsapp.net.
Additionally, the text of the message is'nt in v.Message.GetConversation(), because an web whatsapp message have another structure: 
``
v.Message: extendedTextMessage:{text:"Hi"  contextInfo:{expiration:0  ephemeralSettingTimestamp:0  disappearingMode:{initiator:CHANGED_IN_CHAT}}}  messageContextInfo:{deviceListMetadata:{senderKeyHash:"\xdbn\xed>\xf3$\x1f\xaa\x11["  senderTimestamp:1712857003  recipientKeyHash:"5\xb2\x1c\xb2/\xaf\\\xe6\x023"  recipientTimestamp:1712858566}  deviceListMetadataVersion:2}
``

My code is:

```
func (t *Wigo) onMessage(evt interface{}) {
	switch v := evt.(type) {
	case *events.Message:

		if !v.Info.IsFromMe && !v.Info.IsIncomingBroadcast() && v.Info.Type == "text" && v.Info.Timestamp.After(t.startTime) {

			// Parseamos el sender e idetificamos si tiene el devicepart
			JIDString := v.Info.Sender.ToNonAD().String()
			arrayFromSplit := strings.Split(JIDString, "@")
			numeroTelefono := arrayFromSplit[0]

			if strings.Contains(JIDString, ":") {

			}

			//now := time.Now()
			//t.WAClient.MarkRead([v.Message.GetChat().Id], )

			if t.debug {
				fmt.Println("[WIGO] **************************************************")
				fmt.Println("[WIGO] **************************************************")
				fmt.Println("[WIGO] **************************************************")
				fmt.Println("[WIGO] **************************************************")
				fmt.Println("[WIGO] **************************************************")
				fmt.Println("[WIGO] Recibiste un mensaje: ", v.Message.GetConversation())
				fmt.Println("[WIGO] Tipo de mensaje", v.Info.Type)
				fmt.Println("[WIGO] Sender: ", v.Info.MessageSource.Sender)
				fmt.Printf("[WIGO] Numero de teléfono: %v\n", numeroTelefono)
				fmt.Printf("[WIGO] v.Info: %v\n", v.Info)
				fmt.Printf("[WIGO] v.Message: %v\n", v.Message)
				fmt.Printf("[WIGO] v.Conversation: %v\n", v.Message.GetConversation())
				fmt.Printf("[WIGO] Timestamp: %v\n", v.Info.Timestamp)
			}

			if !env.Get().TestingMode || numeroTelefono == env.Get().TestingNumber {

				cadena, err := t.onMessageChannel(numeroTelefono, v.Message.GetConversation())
				if err != nil {
					panic(err)
				}

				if t.debug {
					fmt.Printf("[WIGO] Enviado respuesta: %v\n", cadena)
				}

				response, err := t.WAClient.SendMessage(context.Background(), v.Info.Sender, &waProto.Message{
					Conversation: proto.String(cadena),
				})
				if err != nil {
					panic(err)
				}

				fmt.Println("[WIGO] Respuesta enviada id: ", response.ID)
			}

		}

	}
}
```

Is there another way to send messages received from whatsapp web? It's behavior is normal?

Thanks!
Ricardo

**Comment by qustavo** [comment]
I had the same problem and ended up doing this:
```go
msg := t.Message.GetConversation()
if msg == "" {
	if x := t.Message.GetExtendedTextMessage(); x != nil {
		msg = *x.Text
	}
}
```
Basically it will get the conversation and fallback to ExtendedTextMessage. I'm not sure why there is this distinction though

**Reply by ghost** [reply]
`Conversation` is just a pointer string for a simple text message.

`ExtendedTextMessage` is a featured text message like quoting, link preview, chat with a disappearing timer, etc.

a chat with disappearing time would always be using `ExtendeTextMessage` for a simple text message.

ref :
https://github.com/tulir/whatsmeow/blob/350073db195c6b6fae832a5b420b5b67fd39f6be/binary/proto/def.pb.go#L7509-L7538
